torch/csrc/aot_inductor/aten_shim_common.cpp

```diff
@@ -0,0 +1,492 @@
+  // Later the statically planned aot_inductor_free will call the original
+  // deleter to free the device storage. We also keep track of those storages in
+  // the debug mode to make sure there is no storage leak.
+  at::Tensor* t = static_cast<at::Tensor*>(aten_tensor);
+  void* data_ptr = t->data_ptr();
+  c10::DataPtr& mutable_ptr =
+      t->storage().unsafeGetStorageImpl()->mutable_data_ptr();
+  DeleterFnPtr deleter = mutable_ptr.get_deleter();
+  AOT_INDUCTOR_CHECK(
+      mutable_ptr.get() == mutable_ptr.get_context() ||
+          deleter == deleteNothing,
+      "Expect data_ and context_ to be the same in a UniqueVoidPtr if the aten \
+        tensor is created by aten; otherwise the aten tensor should be created \
+        by aot_inductor and have a deleteNothing deleter")
+
+  if (deleter != deleteNothing) {
+    bool success = mutable_ptr.compare_exchange_deleter(deleter, deleteNothing);
+    AOT_INDUCTOR_CHECK(
+        success,
+        "Unexpected deleter function on storage, could not swap function");
+    (void)success; // get around the unused variable warning
+  }
+
+  AOTInductorTensorHandle aot_tensor = register_tensor(
+      data_ptr,
+      (void*)deleter,
+      t->dim(),
+      sizes, // sizes has to be passed in because t->sizes() can released early
+      strides, //
+      t->storage_offset(),
+      static_cast<AOTInductorDeviceType>(t->device().type()),
+      static_cast<AOTInductorDeviceIndex>(t->device().index()),
+      static_cast<AOTInductorScalarType>(t->scalar_type()));
+
+#ifndef NDEBUG
+  live_buffer_->insert(static_cast<AOTInductorTensor*>(aot_tensor)->data_ptr);
+#endif
+  return aot_tensor;
+}
+
+void aot_tensor_to_aten_tensor(
+    AOTInductorTensorHandle aot_tensor,
+    AtenTensorHandle aten_tensor) {
+  // Turning AOTInductorTensor into at::Tensor is needed when we want to call
+  // into an aten op as fallback. As described in aten_tensor_to_aot_tensor,
+  // the device storage is managed by AOTInductor's static memory planner.
+  AOTInductorTensor* t = static_cast<AOTInductorTensor*>(aot_tensor);
+  c10::Device device{
+      static_cast<c10::DeviceType>(t->device_type),
+      static_cast<c10::DeviceIndex>(t->device_index)};
+  c10::TensorOptions options = c10::TensorOptions().device(device).dtype(
+      static_cast<c10::ScalarType>(t->dtype));
+
+  // set the deleter to deleteNothing to stop at::Tensor from freeing the
+  // device storage
+  (*static_cast<at::Tensor*>(aten_tensor)) =
+      at::for_blob(t->data_ptr, c10::IntArrayRef(t->sizes, t->ndim))
+          .strides(c10::IntArrayRef(t->strides, t->ndim))
+          .storage_offset(t->storage_offset)
+          .context(t->data_ptr, deleteNothing)
+          .options(options)
+          .make_tensor();
+}
+
+} // namespace
+
+void aot_inductor_initialize() {
+  // c10 types and enums should rarely change, but use this checking to
+  // fail loudly in the worst case
+  AOT_INDUCTOR_CHECK(
```

Comment:
You are in runtime, you have access to all ATen headers, there is no need to use a completely unrelated error checking macro?

That being said, you have sort of kicked up a beehive here. The problem is that it is not defined to raise exceptions across a C ABI boundary. See also https://stackoverflow.com/questions/6426835/what-happens-if-exception-gets-thrown-through-c-code The frame unwinding mechanism is particularly delicate and it would be much more obviously safe if we avoid throwing exceptions this way. To do this, we must explicitly return error codes when an exception would have been raised from this file, and then the model must propagate this exception in some way. The simplest implementation would lose all exception information, but since you will eventually return the runtime, the full error information could likely be stored in a thread local. If it is a burden on codegen to explicitly check all of the error codes, you could choose to turn an error code into an exception inside the model.so (but once again, once you get back to the runtime-model interface, you need to convert back into a C error again.)